// @aztec/stdlib/dest/p2p/gossipable.js
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
export class P2PMessage {
    payload;
    constructor(payload){
        this.payload = payload;
    }
    static fromGossipable(message) {
        return new P2PMessage(message.toBuffer());
    }
    static fromMessageData(messageData) {
        const reader = new BufferReader(messageData);
        const payload = reader.readBuffer();
        return new P2PMessage(payload);
    }
    toMessageData() {
        return serializeToBuffer([
            serializeToBuffer(this.payload.length, this.payload)
        ]);
    }
}
/**
 * Gossipable
 *
 * Any class which extends gossipable will be able to be Gossiped over the p2p network
 */ export class Gossipable {
    cachedId;
    /** The p2p topic identifier, this determines how the message is handled */ static p2pTopic;
    /**
   * A digest of the message information **used for logging only**.
   * The identifier used for deduplication is `getMsgIdFn` as defined in `encoding.ts` which is a hash over topic and data.
   */ async p2pMessageLoggingIdentifier() {
        if (this.cachedId) {
            return this.cachedId;
        }
        this.cachedId = await this.generateP2PMessageIdentifier();
        return this.cachedId;
    }
    toMessage() {
        return this.toBuffer();
    }
}
