// @aztec/accounts/dest/ecdsa/ecdsa_k/index.js
/**
 * The `@aztec/accounts/ecdsa` export provides an ECDSA account contract implementation, that uses an ECDSA private key for authentication, and a Grumpkin key for encryption.
 * Consider using this account type when working with integrations with Ethereum wallets.
 *
 * @packageDocumentation
 */ import { AccountManager } from '@aztec/aztec.js/account';
import { getWallet } from '@aztec/aztec.js/wallet';
import { loadContractArtifact } from '@aztec/stdlib/abi';
import EcdsaKAccountContractJson from '../../../artifacts/EcdsaKAccount.json' with {
    type: 'json'
};
import { EcdsaKBaseAccountContract } from './account_contract.js';
export const EcdsaKAccountContractArtifact = loadContractArtifact(EcdsaKAccountContractJson);
/**
 * Account contract that authenticates transactions using ECDSA signatures
 * verified against a secp256k1 public key stored in an immutable encrypted note.
 * Eagerly loads the contract artifact
 */ export class EcdsaKAccountContract extends EcdsaKBaseAccountContract {
    constructor(signingPrivateKey){
        super(signingPrivateKey);
    }
    getContractArtifact() {
        return Promise.resolve(EcdsaKAccountContractArtifact);
    }
}
/**
 * Creates an Account that relies on an ECDSA signing key for authentication.
 * @param pxe - An PXE server instance.
 * @param secretKey - Secret key used to derive all the keystore keys.
 * @param signingPrivateKey - Secp256k1 key used for signing transactions.
 * @param salt - Deployment salt.
 * @returns An account manager initialized with the account contract and its deployment params
 */ export function getEcdsaKAccount(pxe, secretKey, signingPrivateKey, salt) {
    return AccountManager.create(pxe, secretKey, new EcdsaKAccountContract(signingPrivateKey), salt);
}
/**
 * Gets a wallet for an already registered account using ECDSA signatures.
 * @param pxe - An PXE server instance.
 * @param address - Address for the account.
 * @param signingPrivateKey - ECDSA key used for signing transactions.
 * @returns A wallet for this account that can be used to interact with a contract instance.
 */ export function getEcdsaKWallet(pxe, address, signingPrivateKey) {
    return getWallet(pxe, address, new EcdsaKAccountContract(signingPrivateKey));
}
