// @aztec/bb.js/dest/node-cjs/log/node/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logOptions = void 0;
exports.initLogger = initLogger;
exports.createDebugLogger = createDebugLogger;
const pino_1 = require("pino");
const defaultOptions = {
    name: 'bb.js',
    useOnlyCustomLevels: false,
    customLevels: { verbose: 25 },
};
const defaultLevel = parseLogLevel(process.env.LOG_LEVEL);
let logger;
function initLogger({ level = defaultLevel, useStdErr = false } = { level: defaultLevel, useStdErr: false }) {
    if (logger) {
        return logger;
    }
    exports.logOptions = { level, useStdErr };
    const transport = pino_1.pino.transport({
        target: 'pino/file',
        options: { destination: useStdErr ? 2 : 1 },
    });
    logger = (0, pino_1.pino)({ ...defaultOptions, level }, transport);
}
function createDebugLogger(name) {
    initLogger();
    const sublogger = logger.child({
        name,
    });
    return (msg) => {
        sublogger.debug(msg);
    };
}
function parseLogLevel(logLevel) {
    if (!logLevel) {
        return 'info';
    }
    const knownLogLevels = ['info', 'debug', 'warn', 'error', 'trace', 'silent', 'verbose'];
    const [defaultLogLevel] = logLevel.split(';');
    return knownLogLevels.indexOf(defaultLogLevel) !== -1 ? defaultLogLevel : 'info';
}
