// @aztec/accounts/dest/defaults/account_interface.js
import { DefaultAccountEntrypoint } from '@aztec/entrypoints/account';
import { Fr } from '@aztec/foundation/fields';
/**
 * Default implementation for an account interface. Requires that the account uses the default
 * entrypoint signature, which accept an AppPayload and a FeePayload as defined in noir-libs/aztec-noir/src/entrypoint module
 */ export class DefaultAccountInterface {
    authWitnessProvider;
    address;
    entrypoint;
    chainId;
    version;
    constructor(authWitnessProvider, address, nodeInfo){
        this.authWitnessProvider = authWitnessProvider;
        this.address = address;
        this.entrypoint = new DefaultAccountEntrypoint(address.address, authWitnessProvider, nodeInfo.l1ChainId, nodeInfo.rollupVersion);
        this.chainId = new Fr(nodeInfo.l1ChainId);
        this.version = new Fr(nodeInfo.rollupVersion);
    }
    createTxExecutionRequest(exec, fee, options) {
        return this.entrypoint.createTxExecutionRequest(exec, fee, options);
    }
    createAuthWit(messageHash) {
        return this.authWitnessProvider.createAuthWit(messageHash);
    }
    getCompleteAddress() {
        return this.address;
    }
    getAddress() {
        return this.address.address;
    }
    getChainId() {
        return this.chainId;
    }
    getVersion() {
        return this.version;
    }
}
