// @aztec/protocol-contracts/src/protocol_contract_data.ts

    // GENERATED FILE - DO NOT EDIT. RUN `yarn generate` or `yarn generate:data`
    import { Fr } from '@aztec/foundation/fields';
    import { AztecAddress } from '@aztec/stdlib/aztec-address';

    
    export const protocolContractNames = [
      'AuthRegistry',
'ContractInstanceRegistry',
'ContractClassRegistry',
'MultiCallEntrypoint',
'FeeJuice',
'Router'
    ] as const;

    export type ProtocolContractName = typeof protocolContractNames[number];
  

    
    export const ProtocolContractSalt: Record<ProtocolContractName, Fr> = {
      AuthRegistry: new Fr(1),
ContractInstanceRegistry: new Fr(1),
ContractClassRegistry: new Fr(1),
MultiCallEntrypoint: new Fr(1),
FeeJuice: new Fr(1),
Router: new Fr(1)
    };
  

    
    export const ProtocolContractAddress: Record<ProtocolContractName, AztecAddress> = {
      AuthRegistry: AztecAddress.fromBigInt(1n),
ContractInstanceRegistry: AztecAddress.fromBigInt(2n),
ContractClassRegistry: AztecAddress.fromBigInt(3n),
MultiCallEntrypoint: AztecAddress.fromBigInt(4n),
FeeJuice: AztecAddress.fromBigInt(5n),
Router: AztecAddress.fromBigInt(6n)
    };
  

    
    export const ProtocolContractLeaves = {
      AuthRegistry: Fr.fromHexString('0x1c2474a77af3c756895867fa7622e5e9295121b35849e965f4614d51f00c62e2'),
ContractInstanceRegistry: Fr.fromHexString('0x001e627e78063a4944f1331b4c59544f8ad6d6d478f2fc79b025e8240b94b816'),
ContractClassRegistry: Fr.fromHexString('0x1bd2e13b35e2745e638132262e6642a86ac4219f1b7a95086bebe250f5d557ea'),
MultiCallEntrypoint: Fr.fromHexString('0x06b9a7dd46dc78207ba786ed3776af46ec8a75de544701b8c5852ea69caec77e'),
FeeJuice: Fr.fromHexString('0x2ee34a2d304636f81114dffc8a2357e03e08345a008262eb36112f95aaa58eb1'),
Router: Fr.fromHexString('0x180b58198cce7c0866a828841df6eca6369a71a6e884ca5cf08938301d7ad0ca')
    };
  

    
    export const protocolContractTreeRoot = Fr.fromHexString('0x12e9aa367b065eff3e48912b8cae62209970117d34a8c9ef1e9e4116e41bc8d6');
  

    
  export const CONTRACT_INSTANCE_PUBLISHED_EVENT_TAG = Fr.fromHexString('0x28da3fc471387d4e20b646b77abc8cdfc02a339a7507285a73d84f9b97b13486');
