// @aztec/stdlib/dest/interfaces/configs.js
import { z } from 'zod';
import { schemas } from '../schemas/index.js';
import { AllowedElementSchema } from './allowed_element.js';
export const SequencerConfigSchema = z.object({
    transactionPollingIntervalMS: z.number().optional(),
    maxTxsPerBlock: z.number().optional(),
    minTxsPerBlock: z.number().optional(),
    maxL2BlockGas: z.number().optional(),
    publishTxsWithProposals: z.boolean().optional(),
    maxDABlockGas: z.number().optional(),
    coinbase: schemas.EthAddress.optional(),
    feeRecipient: schemas.AztecAddress.optional(),
    acvmWorkingDirectory: z.string().optional(),
    acvmBinaryPath: z.string().optional(),
    txPublicSetupAllowList: z.array(AllowedElementSchema).optional(),
    maxBlockSizeInBytes: z.number().optional(),
    governanceProposerPayload: schemas.EthAddress.optional(),
    maxL1TxInclusionTimeIntoSlot: z.number().optional(),
    enforceTimeTable: z.boolean().optional(),
    fakeProcessingDelayPerTxMs: z.number().optional(),
    attestationPropagationTime: z.number().optional(),
    skipCollectingAttestations: z.boolean().optional(),
    secondsBeforeInvalidatingBlockAsCommitteeMember: z.number(),
    secondsBeforeInvalidatingBlockAsNonCommitteeMember: z.number(),
    injectFakeAttestation: z.boolean().optional(),
    fishermanMode: z.boolean().optional(),
    shuffleAttestationOrdering: z.boolean().optional()
});
