// @aztec/stdlib/dest/kernel/private_to_public_kernel_circuit_public_inputs.js
import { GeneratorIndex, PRIVATE_TO_PUBLIC_KERNEL_CIRCUIT_PUBLIC_INPUTS_LENGTH } from '@aztec/constants';
import { poseidon2HashWithSeparator } from '@aztec/foundation/crypto';
import { BufferReader, bigintToUInt64BE, serializeToBuffer, serializeToFields } from '@aztec/foundation/serialize';
import { bufferToHex, hexToBuffer } from '@aztec/foundation/string';
import { AztecAddress } from '../aztec-address/index.js';
import { Gas } from '../gas/gas.js';
import { TxConstantData } from '../tx/tx_constant_data.js';
import { PrivateToPublicAccumulatedData } from './private_to_public_accumulated_data.js';
import { PublicCallRequest } from './public_call_request.js';
export class PrivateToPublicKernelCircuitPublicInputs {
    constants;
    nonRevertibleAccumulatedData;
    revertibleAccumulatedData;
    publicTeardownCallRequest;
    gasUsed;
    feePayer;
    includeByTimestamp;
    constructor(constants, nonRevertibleAccumulatedData, revertibleAccumulatedData, publicTeardownCallRequest, gasUsed, feePayer, includeByTimestamp){
        this.constants = constants;
        this.nonRevertibleAccumulatedData = nonRevertibleAccumulatedData;
        this.revertibleAccumulatedData = revertibleAccumulatedData;
        this.publicTeardownCallRequest = publicTeardownCallRequest;
        this.gasUsed = gasUsed;
        this.feePayer = feePayer;
        this.includeByTimestamp = includeByTimestamp;
    }
    toBuffer() {
        return serializeToBuffer(this.constants, this.nonRevertibleAccumulatedData, this.revertibleAccumulatedData, this.publicTeardownCallRequest, this.gasUsed, this.feePayer, bigintToUInt64BE(this.includeByTimestamp));
    }
    static getFields(fields) {
        return [
            fields.constants,
            fields.nonRevertibleAccumulatedData,
            fields.revertibleAccumulatedData,
            fields.publicTeardownCallRequest,
            fields.gasUsed,
            fields.feePayer,
            fields.includeByTimestamp
        ];
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PrivateToPublicKernelCircuitPublicInputs(reader.readObject(TxConstantData), reader.readObject(PrivateToPublicAccumulatedData), reader.readObject(PrivateToPublicAccumulatedData), reader.readObject(PublicCallRequest), reader.readObject(Gas), reader.readObject(AztecAddress), reader.readUInt64());
    }
    static empty() {
        return new PrivateToPublicKernelCircuitPublicInputs(TxConstantData.empty(), PrivateToPublicAccumulatedData.empty(), PrivateToPublicAccumulatedData.empty(), PublicCallRequest.empty(), Gas.empty(), AztecAddress.ZERO, 0n);
    }
    static fromString(str) {
        return PrivateToPublicKernelCircuitPublicInputs.fromBuffer(hexToBuffer(str));
    }
    toString() {
        return bufferToHex(this.toBuffer());
    }
    toFields() {
        const fields = serializeToFields(...PrivateToPublicKernelCircuitPublicInputs.getFields(this));
        if (fields.length !== PRIVATE_TO_PUBLIC_KERNEL_CIRCUIT_PUBLIC_INPUTS_LENGTH) {
            throw new Error(`Invalid number of fields for PrivateToPublicKernelCircuitPublicInputs. Expected ${PRIVATE_TO_PUBLIC_KERNEL_CIRCUIT_PUBLIC_INPUTS_LENGTH}, got ${fields.length}`);
        }
        return fields;
    }
    hash() {
        return poseidon2HashWithSeparator(this.toFields(), GeneratorIndex.PUBLIC_TX_HASH);
    }
}
