// @aztec/bb.js/dest/node/log/node/index.js
import { pino } from 'pino';
const defaultOptions = {
    name: 'bb.js',
    useOnlyCustomLevels: false,
    customLevels: { verbose: 25 },
};
const defaultLevel = parseLogLevel(process.env.LOG_LEVEL);
// Options must be exposed so they can be provided to threads upon creation
// This way we ensure all loggers are spawned with the same options
export let logOptions;
let logger;
export function initLogger({ level = defaultLevel, useStdErr = false } = { level: defaultLevel, useStdErr: false }) {
    if (logger) {
        return logger;
    }
    logOptions = { level, useStdErr };
    const transport = pino.transport({
        target: 'pino/file',
        options: { destination: useStdErr ? 2 : 1 },
    });
    logger = pino({ ...defaultOptions, level }, transport);
}
export function createDebugLogger(name) {
    initLogger();
    const sublogger = logger.child({
        name,
    });
    return (msg) => {
        sublogger.debug(msg);
    };
}
function parseLogLevel(logLevel) {
    if (!logLevel) {
        return 'info';
    }
    const knownLogLevels = ['info', 'debug', 'warn', 'error', 'trace', 'silent', 'verbose'];
    const [defaultLogLevel] = logLevel.split(';');
    return knownLogLevels.indexOf(defaultLogLevel) !== -1 ? defaultLogLevel : 'info';
}
