// @aztec/accounts/dest/copy_cat/base.js
import { AccountWalletWithSecretKey, Fr } from '@aztec/aztec.js';
/**
 * An AccountWallet that copies the address of another account, and then
 * uses the simulation overrides feature to execute different contract code under
 * the copied address. This is used to bypass authwit verification entirely
 * (`is_valid` always returns `true`)
 */ export class CopyCatAccountWalletBase extends AccountWalletWithSecretKey {
    originalAddress;
    artifact;
    instance;
    constructor(pxe, account, originalAddress, artifact, instance){
        super(pxe, account, Fr.ZERO, Fr.ZERO), this.originalAddress = originalAddress, this.artifact = artifact, this.instance = instance;
    }
    getCompleteAddress() {
        return this.originalAddress;
    }
    simulateTx(txRequest, simulatePublic, _skipTxValidation, _skipFeeEnforcement, _overrides) {
        const contractOverrides = {
            [this.originalAddress.address.toString()]: {
                instance: this.instance,
                artifact: this.artifact
            }
        };
        return this.pxe.simulateTx(txRequest, simulatePublic, true, true, {
            contracts: contractOverrides
        });
    }
}
