// @aztec/noir-contracts.js/dest/AMM.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { Contract, ContractBase, DeployMethod, Fr, loadContractArtifact, loadContractArtifactForPublic, PublicKeys } from '@aztec/aztec.js';
import AMMContractArtifactJson from '../artifacts/amm_contract-AMM.json' with {
    type: 'json'
};
export const AMMContractArtifact = loadContractArtifact(AMMContractArtifactJson);
/**
 * Type-safe interface for contract AMM;
 */ export class AMMContract extends ContractBase {
    constructor(instance, wallet){
        super(instance, AMMContractArtifact, wallet);
    }
    /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */ static async at(address, wallet) {
        return Contract.at(address, AMMContract.artifact, wallet);
    }
    /**
   * Creates a tx to deploy a new instance of this contract.
   */ static deploy(wallet, token0, token1, liquidity_token) {
        return new DeployMethod(PublicKeys.default(), wallet, AMMContractArtifact, AMMContract.at, Array.from(arguments).slice(1));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */ static deployWithPublicKeys(publicKeys, wallet, token0, token1, liquidity_token) {
        return new DeployMethod(publicKeys, wallet, AMMContractArtifact, AMMContract.at, Array.from(arguments).slice(2));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */ static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, AMMContractArtifact, AMMContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
   * Returns this contract's artifact.
   */ static get artifact() {
        return AMMContractArtifact;
    }
    /**
   * Returns this contract's artifact with public bytecode.
   */ static get artifactForPublic() {
        return loadContractArtifactForPublic(AMMContractArtifactJson);
    }
    static get storage() {
        return {
            config: {
                slot: new Fr(1n)
            }
        };
    }
}
