// @aztec/noir-contracts.js/dest/AuthRegistry.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { Contract, ContractBase, DeployMethod, Fr, loadContractArtifact, loadContractArtifactForPublic, PublicKeys } from '@aztec/aztec.js';
import AuthRegistryContractArtifactJson from '../artifacts/auth_registry_contract-AuthRegistry.json' with {
    type: 'json'
};
export const AuthRegistryContractArtifact = loadContractArtifact(AuthRegistryContractArtifactJson);
/**
 * Type-safe interface for contract AuthRegistry;
 */ export class AuthRegistryContract extends ContractBase {
    constructor(instance, wallet){
        super(instance, AuthRegistryContractArtifact, wallet);
    }
    /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */ static async at(address, wallet) {
        return Contract.at(address, AuthRegistryContract.artifact, wallet);
    }
    /**
   * Creates a tx to deploy a new instance of this contract.
   */ static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, AuthRegistryContractArtifact, AuthRegistryContract.at, Array.from(arguments).slice(1));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */ static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, AuthRegistryContractArtifact, AuthRegistryContract.at, Array.from(arguments).slice(2));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */ static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, AuthRegistryContractArtifact, AuthRegistryContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
   * Returns this contract's artifact.
   */ static get artifact() {
        return AuthRegistryContractArtifact;
    }
    /**
   * Returns this contract's artifact with public bytecode.
   */ static get artifactForPublic() {
        return loadContractArtifactForPublic(AuthRegistryContractArtifactJson);
    }
    static get storage() {
        return {
            reject_all: {
                slot: new Fr(1n)
            },
            approved_actions: {
                slot: new Fr(2n)
            }
        };
    }
}
