// @aztec/noir-contracts.js/dest/Crowdfunding.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { Contract, ContractBase, DeployMethod, EventSelector, Fr, loadContractArtifact, loadContractArtifactForPublic, PublicKeys } from '@aztec/aztec.js';
import CrowdfundingContractArtifactJson from '../artifacts/crowdfunding_contract-Crowdfunding.json' with {
    type: 'json'
};
export const CrowdfundingContractArtifact = loadContractArtifact(CrowdfundingContractArtifactJson);
/**
 * Type-safe interface for contract Crowdfunding;
 */ export class CrowdfundingContract extends ContractBase {
    constructor(instance, wallet){
        super(instance, CrowdfundingContractArtifact, wallet);
    }
    /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */ static async at(address, wallet) {
        return Contract.at(address, CrowdfundingContract.artifact, wallet);
    }
    /**
   * Creates a tx to deploy a new instance of this contract.
   */ static deploy(wallet, donation_token, operator, deadline) {
        return new DeployMethod(PublicKeys.default(), wallet, CrowdfundingContractArtifact, CrowdfundingContract.at, Array.from(arguments).slice(1));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */ static deployWithPublicKeys(publicKeys, wallet, donation_token, operator, deadline) {
        return new DeployMethod(publicKeys, wallet, CrowdfundingContractArtifact, CrowdfundingContract.at, Array.from(arguments).slice(2));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */ static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, CrowdfundingContractArtifact, CrowdfundingContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
   * Returns this contract's artifact.
   */ static get artifact() {
        return CrowdfundingContractArtifact;
    }
    /**
   * Returns this contract's artifact with public bytecode.
   */ static get artifactForPublic() {
        return loadContractArtifactForPublic(CrowdfundingContractArtifactJson);
    }
    static get storage() {
        return {
            config: {
                slot: new Fr(1n)
            },
            donation_receipts: {
                slot: new Fr(5n)
            }
        };
    }
    static get events() {
        return {
            WithdrawalProcessed: {
                abiType: {
                    "kind": "struct",
                    "fields": [
                        {
                            "name": "who",
                            "type": {
                                "kind": "struct",
                                "fields": [
                                    {
                                        "name": "inner",
                                        "type": {
                                            "kind": "field"
                                        }
                                    }
                                ],
                                "path": "aztec::protocol_types::address::aztec_address::AztecAddress"
                            }
                        },
                        {
                            "name": "amount",
                            "type": {
                                "kind": "integer",
                                "sign": "unsigned",
                                "width": 128
                            }
                        }
                    ],
                    "path": "Crowdfunding::WithdrawalProcessed"
                },
                eventSelector: EventSelector.fromString("0x6f9303c1"),
                fieldNames: [
                    "who",
                    "amount"
                ]
            }
        };
    }
}
