// @aztec/noir-contracts.js/dest/EcdsaRAccount.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { Contract, ContractBase, DeployMethod, Fr, loadContractArtifact, loadContractArtifactForPublic, PublicKeys } from '@aztec/aztec.js';
import EcdsaRAccountContractArtifactJson from '../artifacts/ecdsa_r_account_contract-EcdsaRAccount.json' with {
    type: 'json'
};
export const EcdsaRAccountContractArtifact = loadContractArtifact(EcdsaRAccountContractArtifactJson);
/**
 * Type-safe interface for contract EcdsaRAccount;
 */ export class EcdsaRAccountContract extends ContractBase {
    constructor(instance, wallet){
        super(instance, EcdsaRAccountContractArtifact, wallet);
    }
    /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */ static async at(address, wallet) {
        return Contract.at(address, EcdsaRAccountContract.artifact, wallet);
    }
    /**
   * Creates a tx to deploy a new instance of this contract.
   */ static deploy(wallet, signing_pub_key_x, signing_pub_key_y) {
        return new DeployMethod(PublicKeys.default(), wallet, EcdsaRAccountContractArtifact, EcdsaRAccountContract.at, Array.from(arguments).slice(1));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */ static deployWithPublicKeys(publicKeys, wallet, signing_pub_key_x, signing_pub_key_y) {
        return new DeployMethod(publicKeys, wallet, EcdsaRAccountContractArtifact, EcdsaRAccountContract.at, Array.from(arguments).slice(2));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */ static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, EcdsaRAccountContractArtifact, EcdsaRAccountContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
   * Returns this contract's artifact.
   */ static get artifact() {
        return EcdsaRAccountContractArtifact;
    }
    /**
   * Returns this contract's artifact with public bytecode.
   */ static get artifactForPublic() {
        return loadContractArtifactForPublic(EcdsaRAccountContractArtifactJson);
    }
    static get storage() {
        return {
            signing_public_key: {
                slot: new Fr(1n)
            }
        };
    }
}
