// @aztec/noir-contracts.js/dest/FeeJuice.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { Contract, ContractBase, DeployMethod, Fr, loadContractArtifact, loadContractArtifactForPublic, PublicKeys } from '@aztec/aztec.js';
import FeeJuiceContractArtifactJson from '../artifacts/fee_juice_contract-FeeJuice.json' with {
    type: 'json'
};
export const FeeJuiceContractArtifact = loadContractArtifact(FeeJuiceContractArtifactJson);
/**
 * Type-safe interface for contract FeeJuice;
 */ export class FeeJuiceContract extends ContractBase {
    constructor(instance, wallet){
        super(instance, FeeJuiceContractArtifact, wallet);
    }
    /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */ static async at(address, wallet) {
        return Contract.at(address, FeeJuiceContract.artifact, wallet);
    }
    /**
   * Creates a tx to deploy a new instance of this contract.
   */ static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, FeeJuiceContractArtifact, FeeJuiceContract.at, Array.from(arguments).slice(1));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */ static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, FeeJuiceContractArtifact, FeeJuiceContract.at, Array.from(arguments).slice(2));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */ static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, FeeJuiceContractArtifact, FeeJuiceContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
   * Returns this contract's artifact.
   */ static get artifact() {
        return FeeJuiceContractArtifact;
    }
    /**
   * Returns this contract's artifact with public bytecode.
   */ static get artifactForPublic() {
        return loadContractArtifactForPublic(FeeJuiceContractArtifactJson);
    }
    static get storage() {
        return {
            balances: {
                slot: new Fr(1n)
            }
        };
    }
}
