// @aztec/noir-contracts.js/dest/NFT.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { Contract, ContractBase, DeployMethod, Fr, loadContractArtifact, loadContractArtifactForPublic, PublicKeys } from '@aztec/aztec.js';
import NFTContractArtifactJson from '../artifacts/nft_contract-NFT.json' with {
    type: 'json'
};
export const NFTContractArtifact = loadContractArtifact(NFTContractArtifactJson);
/**
 * Type-safe interface for contract NFT;
 */ export class NFTContract extends ContractBase {
    constructor(instance, wallet){
        super(instance, NFTContractArtifact, wallet);
    }
    /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */ static async at(address, wallet) {
        return Contract.at(address, NFTContract.artifact, wallet);
    }
    /**
   * Creates a tx to deploy a new instance of this contract.
   */ static deploy(wallet, admin, name, symbol) {
        return new DeployMethod(PublicKeys.default(), wallet, NFTContractArtifact, NFTContract.at, Array.from(arguments).slice(1));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */ static deployWithPublicKeys(publicKeys, wallet, admin, name, symbol) {
        return new DeployMethod(publicKeys, wallet, NFTContractArtifact, NFTContract.at, Array.from(arguments).slice(2));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */ static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, NFTContractArtifact, NFTContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
   * Returns this contract's artifact.
   */ static get artifact() {
        return NFTContractArtifact;
    }
    /**
   * Returns this contract's artifact with public bytecode.
   */ static get artifactForPublic() {
        return loadContractArtifactForPublic(NFTContractArtifactJson);
    }
    static get storage() {
        return {
            symbol: {
                slot: new Fr(1n)
            },
            name: {
                slot: new Fr(3n)
            },
            admin: {
                slot: new Fr(5n)
            },
            minters: {
                slot: new Fr(6n)
            },
            private_nfts: {
                slot: new Fr(7n)
            },
            nft_exists: {
                slot: new Fr(8n)
            },
            public_owners: {
                slot: new Fr(9n)
            }
        };
    }
}
