// @aztec/noir-contracts.js/dest/Orderbook.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { Contract, ContractBase, DeployMethod, EventSelector, Fr, loadContractArtifact, loadContractArtifactForPublic, PublicKeys } from '@aztec/aztec.js';
import OrderbookContractArtifactJson from '../artifacts/orderbook_contract-Orderbook.json' with {
    type: 'json'
};
export const OrderbookContractArtifact = loadContractArtifact(OrderbookContractArtifactJson);
/**
 * Type-safe interface for contract Orderbook;
 */ export class OrderbookContract extends ContractBase {
    constructor(instance, wallet){
        super(instance, OrderbookContractArtifact, wallet);
    }
    /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */ static async at(address, wallet) {
        return Contract.at(address, OrderbookContract.artifact, wallet);
    }
    /**
   * Creates a tx to deploy a new instance of this contract.
   */ static deploy(wallet, token0, token1) {
        return new DeployMethod(PublicKeys.default(), wallet, OrderbookContractArtifact, OrderbookContract.at, Array.from(arguments).slice(1));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */ static deployWithPublicKeys(publicKeys, wallet, token0, token1) {
        return new DeployMethod(publicKeys, wallet, OrderbookContractArtifact, OrderbookContract.at, Array.from(arguments).slice(2));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */ static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, OrderbookContractArtifact, OrderbookContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
   * Returns this contract's artifact.
   */ static get artifact() {
        return OrderbookContractArtifact;
    }
    /**
   * Returns this contract's artifact with public bytecode.
   */ static get artifactForPublic() {
        return loadContractArtifactForPublic(OrderbookContractArtifactJson);
    }
    static get storage() {
        return {
            config: {
                slot: new Fr(1n)
            },
            orders: {
                slot: new Fr(4n)
            }
        };
    }
    static get events() {
        return {
            OrderFulfilled: {
                abiType: {
                    "kind": "struct",
                    "fields": [
                        {
                            "name": "order_id",
                            "type": {
                                "kind": "field"
                            }
                        }
                    ],
                    "path": "Orderbook::OrderFulfilled"
                },
                eventSelector: EventSelector.fromString("0x0e2d8a86"),
                fieldNames: [
                    "order_id"
                ]
            },
            OrderCreated: {
                abiType: {
                    "kind": "struct",
                    "fields": [
                        {
                            "name": "order_id",
                            "type": {
                                "kind": "field"
                            }
                        }
                    ],
                    "path": "Orderbook::OrderCreated"
                },
                eventSelector: EventSelector.fromString("0xc341651c"),
                fieldNames: [
                    "order_id"
                ]
            }
        };
    }
}
