// @aztec/noir-contracts.js/dest/PriceFeed.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { Contract, ContractBase, DeployMethod, Fr, loadContractArtifact, loadContractArtifactForPublic, PublicKeys } from '@aztec/aztec.js';
import PriceFeedContractArtifactJson from '../artifacts/price_feed_contract-PriceFeed.json' with {
    type: 'json'
};
export const PriceFeedContractArtifact = loadContractArtifact(PriceFeedContractArtifactJson);
/**
 * Type-safe interface for contract PriceFeed;
 */ export class PriceFeedContract extends ContractBase {
    constructor(instance, wallet){
        super(instance, PriceFeedContractArtifact, wallet);
    }
    /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */ static async at(address, wallet) {
        return Contract.at(address, PriceFeedContract.artifact, wallet);
    }
    /**
   * Creates a tx to deploy a new instance of this contract.
   */ static deploy(wallet) {
        return new DeployMethod(PublicKeys.default(), wallet, PriceFeedContractArtifact, PriceFeedContract.at, Array.from(arguments).slice(1));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */ static deployWithPublicKeys(publicKeys, wallet) {
        return new DeployMethod(publicKeys, wallet, PriceFeedContractArtifact, PriceFeedContract.at, Array.from(arguments).slice(2));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */ static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, PriceFeedContractArtifact, PriceFeedContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
   * Returns this contract's artifact.
   */ static get artifact() {
        return PriceFeedContractArtifact;
    }
    /**
   * Returns this contract's artifact with public bytecode.
   */ static get artifactForPublic() {
        return loadContractArtifactForPublic(PriceFeedContractArtifactJson);
    }
    static get storage() {
        return {
            assets: {
                slot: new Fr(1n)
            }
        };
    }
}
