// @aztec/noir-contracts.js/dest/PrivateVoting.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { Contract, ContractBase, DeployMethod, Fr, loadContractArtifact, loadContractArtifactForPublic, PublicKeys } from '@aztec/aztec.js';
import PrivateVotingContractArtifactJson from '../artifacts/private_voting_contract-PrivateVoting.json' with {
    type: 'json'
};
export const PrivateVotingContractArtifact = loadContractArtifact(PrivateVotingContractArtifactJson);
/**
 * Type-safe interface for contract PrivateVoting;
 */ export class PrivateVotingContract extends ContractBase {
    constructor(instance, wallet){
        super(instance, PrivateVotingContractArtifact, wallet);
    }
    /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */ static async at(address, wallet) {
        return Contract.at(address, PrivateVotingContract.artifact, wallet);
    }
    /**
   * Creates a tx to deploy a new instance of this contract.
   */ static deploy(wallet, admin) {
        return new DeployMethod(PublicKeys.default(), wallet, PrivateVotingContractArtifact, PrivateVotingContract.at, Array.from(arguments).slice(1));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */ static deployWithPublicKeys(publicKeys, wallet, admin) {
        return new DeployMethod(publicKeys, wallet, PrivateVotingContractArtifact, PrivateVotingContract.at, Array.from(arguments).slice(2));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */ static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, PrivateVotingContractArtifact, PrivateVotingContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
   * Returns this contract's artifact.
   */ static get artifact() {
        return PrivateVotingContractArtifact;
    }
    /**
   * Returns this contract's artifact with public bytecode.
   */ static get artifactForPublic() {
        return loadContractArtifactForPublic(PrivateVotingContractArtifactJson);
    }
    static get storage() {
        return {
            admin: {
                slot: new Fr(1n)
            },
            tally: {
                slot: new Fr(2n)
            },
            vote_ended: {
                slot: new Fr(3n)
            },
            active_at_block: {
                slot: new Fr(4n)
            }
        };
    }
}
