// @aztec/noir-contracts.js/dest/PrivateToken.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { Contract, ContractBase, DeployMethod, Fr, loadContractArtifact, loadContractArtifactForPublic, PublicKeys } from '@aztec/aztec.js';
import PrivateTokenContractArtifactJson from '../artifacts/private_token_contract-PrivateToken.json' with {
    type: 'json'
};
export const PrivateTokenContractArtifact = loadContractArtifact(PrivateTokenContractArtifactJson);
/**
 * Type-safe interface for contract PrivateToken;
 */ export class PrivateTokenContract extends ContractBase {
    constructor(instance, wallet){
        super(instance, PrivateTokenContractArtifact, wallet);
    }
    /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */ static async at(address, wallet) {
        return Contract.at(address, PrivateTokenContract.artifact, wallet);
    }
    /**
   * Creates a tx to deploy a new instance of this contract.
   */ static deploy(wallet, initial_supply, owner) {
        return new DeployMethod(PublicKeys.default(), wallet, PrivateTokenContractArtifact, PrivateTokenContract.at, Array.from(arguments).slice(1));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */ static deployWithPublicKeys(publicKeys, wallet, initial_supply, owner) {
        return new DeployMethod(publicKeys, wallet, PrivateTokenContractArtifact, PrivateTokenContract.at, Array.from(arguments).slice(2));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */ static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, PrivateTokenContractArtifact, PrivateTokenContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
   * Returns this contract's artifact.
   */ static get artifact() {
        return PrivateTokenContractArtifact;
    }
    /**
   * Returns this contract's artifact with public bytecode.
   */ static get artifactForPublic() {
        return loadContractArtifactForPublic(PrivateTokenContractArtifactJson);
    }
    static get storage() {
        return {
            balances: {
                slot: new Fr(1n)
            }
        };
    }
}
