// @aztec/noir-contracts.js/dest/Token.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { Contract, ContractBase, DeployMethod, EventSelector, Fr, loadContractArtifact, loadContractArtifactForPublic, PublicKeys } from '@aztec/aztec.js';
import TokenContractArtifactJson from '../artifacts/token_contract-Token.json' with {
    type: 'json'
};
export const TokenContractArtifact = loadContractArtifact(TokenContractArtifactJson);
/**
 * Type-safe interface for contract Token;
 */ export class TokenContract extends ContractBase {
    constructor(instance, wallet){
        super(instance, TokenContractArtifact, wallet);
    }
    /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */ static async at(address, wallet) {
        return Contract.at(address, TokenContract.artifact, wallet);
    }
    /**
   * Creates a tx to deploy a new instance of this contract.
   */ static deploy(wallet, admin, name, symbol, decimals) {
        return new DeployMethod(PublicKeys.default(), wallet, TokenContractArtifact, TokenContract.at, Array.from(arguments).slice(1));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */ static deployWithPublicKeys(publicKeys, wallet, admin, name, symbol, decimals) {
        return new DeployMethod(publicKeys, wallet, TokenContractArtifact, TokenContract.at, Array.from(arguments).slice(2));
    }
    /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */ static deployWithOpts(opts, ...args) {
        return new DeployMethod(opts.publicKeys ?? PublicKeys.default(), opts.wallet, TokenContractArtifact, TokenContract.at, Array.from(arguments).slice(1), opts.method ?? 'constructor');
    }
    /**
   * Returns this contract's artifact.
   */ static get artifact() {
        return TokenContractArtifact;
    }
    /**
   * Returns this contract's artifact with public bytecode.
   */ static get artifactForPublic() {
        return loadContractArtifactForPublic(TokenContractArtifactJson);
    }
    static get storage() {
        return {
            admin: {
                slot: new Fr(1n)
            },
            minters: {
                slot: new Fr(2n)
            },
            balances: {
                slot: new Fr(3n)
            },
            total_supply: {
                slot: new Fr(4n)
            },
            public_balances: {
                slot: new Fr(5n)
            },
            symbol: {
                slot: new Fr(6n)
            },
            name: {
                slot: new Fr(8n)
            },
            decimals: {
                slot: new Fr(10n)
            }
        };
    }
    static get events() {
        return {
            Transfer: {
                abiType: {
                    "kind": "struct",
                    "fields": [
                        {
                            "name": "from",
                            "type": {
                                "kind": "struct",
                                "fields": [
                                    {
                                        "name": "inner",
                                        "type": {
                                            "kind": "field"
                                        }
                                    }
                                ],
                                "path": "aztec::protocol_types::address::aztec_address::AztecAddress"
                            }
                        },
                        {
                            "name": "to",
                            "type": {
                                "kind": "struct",
                                "fields": [
                                    {
                                        "name": "inner",
                                        "type": {
                                            "kind": "field"
                                        }
                                    }
                                ],
                                "path": "aztec::protocol_types::address::aztec_address::AztecAddress"
                            }
                        },
                        {
                            "name": "amount",
                            "type": {
                                "kind": "integer",
                                "sign": "unsigned",
                                "width": 128
                            }
                        }
                    ],
                    "path": "Token::Transfer"
                },
                eventSelector: EventSelector.fromString("0x70a1894e"),
                fieldNames: [
                    "from",
                    "to",
                    "amount"
                ]
            }
        };
    }
}
