// @aztec/stdlib/dest/rollup/block_constant_data.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { AppendOnlyTreeSnapshot } from '../trees/append_only_tree_snapshot.js';
import { GlobalVariables } from '../tx/global_variables.js';
/**
 * Constants that are the same for the entire block.
 */ export class BlockConstantData {
    lastArchive;
    newL1ToL2;
    vkTreeRoot;
    protocolContractTreeRoot;
    globalVariables;
    constructor(/** Archive tree snapshot at the very beginning of the entire rollup. */ lastArchive, /** L1 to L2 message tree snapshot after this block lands. */ newL1ToL2, /** Root of the verification key tree. */ vkTreeRoot, /** Root of the protocol contract tree. */ protocolContractTreeRoot, /** Global variables for the block. */ globalVariables){
        this.lastArchive = lastArchive;
        this.newL1ToL2 = newL1ToL2;
        this.vkTreeRoot = vkTreeRoot;
        this.protocolContractTreeRoot = protocolContractTreeRoot;
        this.globalVariables = globalVariables;
    }
    static from(fields) {
        return new BlockConstantData(...BlockConstantData.getFields(fields));
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new BlockConstantData(reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), Fr.fromBuffer(reader), Fr.fromBuffer(reader), reader.readObject(GlobalVariables));
    }
    static getFields(fields) {
        return [
            fields.lastArchive,
            fields.newL1ToL2,
            fields.vkTreeRoot,
            fields.protocolContractTreeRoot,
            fields.globalVariables
        ];
    }
    static empty() {
        return new BlockConstantData(AppendOnlyTreeSnapshot.empty(), AppendOnlyTreeSnapshot.empty(), Fr.ZERO, Fr.ZERO, GlobalVariables.empty());
    }
    toBuffer() {
        return serializeToBuffer(...BlockConstantData.getFields(this));
    }
}
