// @aztec/stdlib/dest/rollup/epoch_constant_data.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
/**
 * Constants that are the same for the entire epoch.
 */ export class EpochConstantData {
    vkTreeRoot;
    protocolContractTreeRoot;
    proverId;
    constructor(/**
     * Root of the verification key tree.
     */ vkTreeRoot, /**
     * Root of the protocol contract tree.
     */ protocolContractTreeRoot, /**
     * Identifier of the prover of the epoch.
     */ proverId){
        this.vkTreeRoot = vkTreeRoot;
        this.protocolContractTreeRoot = protocolContractTreeRoot;
        this.proverId = proverId;
    }
    static from(fields) {
        return new EpochConstantData(...EpochConstantData.getFields(fields));
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new EpochConstantData(Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader));
    }
    static getFields(fields) {
        return [
            fields.vkTreeRoot,
            fields.protocolContractTreeRoot,
            fields.proverId
        ];
    }
    static empty() {
        return new EpochConstantData(Fr.ZERO, Fr.ZERO, Fr.ZERO);
    }
    toBuffer() {
        return serializeToBuffer(...EpochConstantData.getFields(this));
    }
}
