// @aztec/stdlib/dest/rollup/padding_block_root_rollup_inputs.js
import { bufferSchemaFor } from '@aztec/foundation/schemas';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { bufferToHex, hexToBuffer } from '@aztec/foundation/string';
import { EpochConstantData } from './epoch_constant_data.js';
/**
 * Inputs of the padding block root rollup circuit.
 */ export class PaddingBlockRootRollupInputs {
    constants;
    constructor(constants){
        this.constants = constants;
    }
    static from(fields) {
        return new PaddingBlockRootRollupInputs(...PaddingBlockRootRollupInputs.getFields(fields));
    }
    static getFields(fields) {
        return [
            fields.constants
        ];
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PaddingBlockRootRollupInputs(reader.readObject(EpochConstantData));
    }
    toBuffer() {
        return serializeToBuffer(...PaddingBlockRootRollupInputs.getFields(this));
    }
    static fromString(str) {
        return PaddingBlockRootRollupInputs.fromBuffer(hexToBuffer(str));
    }
    toString() {
        return bufferToHex(this.toBuffer());
    }
    /** Returns a buffer representation for JSON serialization. */ toJSON() {
        return this.toBuffer();
    }
    static get schema() {
        return bufferSchemaFor(PaddingBlockRootRollupInputs);
    }
}
