// @aztec/stdlib/dest/rollup/private_base_rollup_inputs.js
import { bufferSchemaFor } from '@aztec/foundation/schemas';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { bufferToHex, hexToBuffer } from '@aztec/foundation/string';
import { PrivateBaseRollupHints } from './base_rollup_hints.js';
import { PrivateTubeData } from './private_tube_data.js';
export class PrivateBaseRollupInputs {
    tubeData;
    hints;
    constructor(tubeData, hints){
        this.tubeData = tubeData;
        this.hints = hints;
    }
    static from(fields) {
        return new PrivateBaseRollupInputs(...PrivateBaseRollupInputs.getFields(fields));
    }
    static getFields(fields) {
        return [
            fields.tubeData,
            fields.hints
        ];
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PrivateBaseRollupInputs(reader.readObject(PrivateTubeData), reader.readObject(PrivateBaseRollupHints));
    }
    toBuffer() {
        return serializeToBuffer(...PrivateBaseRollupInputs.getFields(this));
    }
    static fromString(str) {
        return PrivateBaseRollupInputs.fromBuffer(hexToBuffer(str));
    }
    toString() {
        return bufferToHex(this.toBuffer());
    }
    static empty() {
        return new PrivateBaseRollupInputs(PrivateTubeData.empty(), PrivateBaseRollupHints.empty());
    }
    /** Returns a buffer representation for JSON serialization. */ toJSON() {
        return this.toBuffer();
    }
    /** Creates an instance from a hex string. */ static get schema() {
        return bufferSchemaFor(PrivateBaseRollupInputs);
    }
}
