// @aztec/stdlib/dest/rollup/public_tube_data.js
import { NESTED_RECURSIVE_ROLLUP_HONK_PROOF_LENGTH } from '@aztec/constants';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { PrivateToPublicKernelCircuitPublicInputs } from '../kernel/private_to_public_kernel_circuit_public_inputs.js';
import { RecursiveProof, makeEmptyRecursiveProof } from '../proofs/recursive_proof.js';
import { VkData } from '../vks/index.js';
export class PublicTubeData {
    publicInputs;
    proof;
    vkData;
    constructor(publicInputs, proof, vkData){
        this.publicInputs = publicInputs;
        this.proof = proof;
        this.vkData = vkData;
    }
    static empty() {
        return new PublicTubeData(PrivateToPublicKernelCircuitPublicInputs.empty(), makeEmptyRecursiveProof(NESTED_RECURSIVE_ROLLUP_HONK_PROOF_LENGTH), VkData.empty());
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PublicTubeData(reader.readObject(PrivateToPublicKernelCircuitPublicInputs), RecursiveProof.fromBuffer(reader, NESTED_RECURSIVE_ROLLUP_HONK_PROOF_LENGTH), reader.readObject(VkData));
    }
    toBuffer() {
        return serializeToBuffer(this.publicInputs, this.proof, this.vkData);
    }
}
