// @aztec/noir-contracts.js/src/AMM.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import AMMContractArtifactJson from '../artifacts/amm_contract-AMM.json' with { type: 'json' };
export const AMMContractArtifact = loadContractArtifact(AMMContractArtifactJson as NoirCompiledContract);



/**
 * Type-safe interface for contract AMM;
 */
export class AMMContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, AMMContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, AMMContract.artifact, wallet) as Promise<AMMContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, token0: AztecAddressLike, token1: AztecAddressLike, liquidity_token: AztecAddressLike) {
    return new DeployMethod<AMMContract>(PublicKeys.default(), wallet, AMMContractArtifact, AMMContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, token0: AztecAddressLike, token1: AztecAddressLike, liquidity_token: AztecAddressLike) {
    return new DeployMethod<AMMContract>(publicKeys, wallet, AMMContractArtifact, AMMContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof AMMContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<AMMContract['methods'][M]>
  ) {
    return new DeployMethod<AMMContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      AMMContractArtifact,
      AMMContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return AMMContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(AMMContractArtifactJson as NoirCompiledContract);
  }
  

  public static get storage(): ContractStorageLayout<'config'> {
      return {
        config: {
      slot: new Fr(1n),
    }
      } as ContractStorageLayout<'config'>;
    }
    

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** add_liquidity(amount0_max: integer, amount1_max: integer, amount0_min: integer, amount1_min: integer, authwit_nonce: field) */
    add_liquidity: ((amount0_max: (bigint | number), amount1_max: (bigint | number), amount0_min: (bigint | number), amount1_min: (bigint | number), authwit_nonce: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** constructor(token0: struct, token1: struct, liquidity_token: struct) */
    constructor: ((token0: AztecAddressLike, token1: AztecAddressLike, liquidity_token: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_amount_in_for_exact_out(balance_in: integer, balance_out: integer, amount_out: integer) */
    get_amount_in_for_exact_out: ((balance_in: (bigint | number), balance_out: (bigint | number), amount_out: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_amount_out_for_exact_in(balance_in: integer, balance_out: integer, amount_in: integer) */
    get_amount_out_for_exact_in: ((balance_in: (bigint | number), balance_out: (bigint | number), amount_in: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** process_message(message_ciphertext: struct, message_context: struct) */
    process_message: ((message_ciphertext: FieldLike[], message_context: { tx_hash: FieldLike, unique_note_hashes_in_tx: FieldLike[], first_nullifier_in_tx: FieldLike, recipient: AztecAddressLike }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** remove_liquidity(liquidity: integer, amount0_min: integer, amount1_min: integer, authwit_nonce: field) */
    remove_liquidity: ((liquidity: (bigint | number), amount0_min: (bigint | number), amount1_min: (bigint | number), authwit_nonce: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** swap_exact_tokens_for_tokens(token_in: struct, token_out: struct, amount_in: integer, amount_out_min: integer, authwit_nonce: field) */
    swap_exact_tokens_for_tokens: ((token_in: AztecAddressLike, token_out: AztecAddressLike, amount_in: (bigint | number), amount_out_min: (bigint | number), authwit_nonce: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** swap_tokens_for_exact_tokens(token_in: struct, token_out: struct, amount_out: integer, amount_in_max: integer, authwit_nonce: field) */
    swap_tokens_for_exact_tokens: ((token_in: AztecAddressLike, token_out: AztecAddressLike, amount_out: (bigint | number), amount_in_max: (bigint | number), authwit_nonce: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_private_state() */
    sync_private_state: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
}
