// @aztec/noir-contracts.js/src/AppSubscription.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import AppSubscriptionContractArtifactJson from '../artifacts/app_subscription_contract-AppSubscription.json' with { type: 'json' };
export const AppSubscriptionContractArtifact = loadContractArtifact(AppSubscriptionContractArtifactJson as NoirCompiledContract);



/**
 * Type-safe interface for contract AppSubscription;
 */
export class AppSubscriptionContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, AppSubscriptionContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, AppSubscriptionContract.artifact, wallet) as Promise<AppSubscriptionContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, target_address: AztecAddressLike, subscription_recipient_address: AztecAddressLike, subscription_token_address: AztecAddressLike, subscription_price: (bigint | number), fee_juice_limit_per_tx: FieldLike) {
    return new DeployMethod<AppSubscriptionContract>(PublicKeys.default(), wallet, AppSubscriptionContractArtifact, AppSubscriptionContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, target_address: AztecAddressLike, subscription_recipient_address: AztecAddressLike, subscription_token_address: AztecAddressLike, subscription_price: (bigint | number), fee_juice_limit_per_tx: FieldLike) {
    return new DeployMethod<AppSubscriptionContract>(publicKeys, wallet, AppSubscriptionContractArtifact, AppSubscriptionContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof AppSubscriptionContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<AppSubscriptionContract['methods'][M]>
  ) {
    return new DeployMethod<AppSubscriptionContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      AppSubscriptionContractArtifact,
      AppSubscriptionContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return AppSubscriptionContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(AppSubscriptionContractArtifactJson as NoirCompiledContract);
  }
  

  public static get storage(): ContractStorageLayout<'config' | 'subscriptions'> {
      return {
        config: {
      slot: new Fr(1n),
    },
subscriptions: {
      slot: new Fr(7n),
    }
      } as ContractStorageLayout<'config' | 'subscriptions'>;
    }
    

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** constructor(target_address: struct, subscription_recipient_address: struct, subscription_token_address: struct, subscription_price: integer, fee_juice_limit_per_tx: field) */
    constructor: ((target_address: AztecAddressLike, subscription_recipient_address: AztecAddressLike, subscription_token_address: AztecAddressLike, subscription_price: (bigint | number), fee_juice_limit_per_tx: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** entrypoint(payload: struct, user_address: struct) */
    entrypoint: ((payload: { function_calls: { args_hash: FieldLike, function_selector: FunctionSelectorLike, target_address: AztecAddressLike, is_public: boolean, is_static: boolean }[], tx_nonce: FieldLike }, user_address: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** is_initialized(subscriber: struct) */
    is_initialized: ((subscriber: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** process_message(message_ciphertext: struct, message_context: struct) */
    process_message: ((message_ciphertext: FieldLike[], message_context: { tx_hash: FieldLike, unique_note_hashes_in_tx: FieldLike[], first_nullifier_in_tx: FieldLike, recipient: AztecAddressLike }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** subscribe(subscriber: struct, authwit_nonce: field, expiry_block_number: integer, tx_count: integer) */
    subscribe: ((subscriber: AztecAddressLike, authwit_nonce: FieldLike, expiry_block_number: (bigint | number), tx_count: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_private_state() */
    sync_private_state: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
}
