// @aztec/noir-contracts.js/src/Auth.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import AuthContractArtifactJson from '../artifacts/auth_contract-Auth.json' with { type: 'json' };
export const AuthContractArtifact = loadContractArtifact(AuthContractArtifactJson as NoirCompiledContract);



/**
 * Type-safe interface for contract Auth;
 */
export class AuthContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, AuthContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, AuthContract.artifact, wallet) as Promise<AuthContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, admin: AztecAddressLike) {
    return new DeployMethod<AuthContract>(PublicKeys.default(), wallet, AuthContractArtifact, AuthContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, admin: AztecAddressLike) {
    return new DeployMethod<AuthContract>(publicKeys, wallet, AuthContractArtifact, AuthContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof AuthContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<AuthContract['methods'][M]>
  ) {
    return new DeployMethod<AuthContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      AuthContractArtifact,
      AuthContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return AuthContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(AuthContractArtifactJson as NoirCompiledContract);
  }
  

  public static get storage(): ContractStorageLayout<'admin' | 'authorized'> {
      return {
        admin: {
      slot: new Fr(1n),
    },
authorized: {
      slot: new Fr(3n),
    }
      } as ContractStorageLayout<'admin' | 'authorized'>;
    }
    

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** constructor(admin: struct) */
    constructor: ((admin: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** do_private_authorized_thing() */
    do_private_authorized_thing: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_authorized() */
    get_authorized: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_authorized_delay() */
    get_authorized_delay: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_authorized_in_private() */
    get_authorized_in_private: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_scheduled_authorized() */
    get_scheduled_authorized: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** process_message(message_ciphertext: struct, message_context: struct) */
    process_message: ((message_ciphertext: FieldLike[], message_context: { tx_hash: FieldLike, unique_note_hashes_in_tx: FieldLike[], first_nullifier_in_tx: FieldLike, recipient: AztecAddressLike }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** set_authorized(authorized: struct) */
    set_authorized: ((authorized: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** set_authorized_delay(new_delay: integer) */
    set_authorized_delay: ((new_delay: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_private_state() */
    sync_private_state: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
}
