// @aztec/noir-contracts.js/src/CardGame.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import CardGameContractArtifactJson from '../artifacts/card_game_contract-CardGame.json' with { type: 'json' };
export const CardGameContractArtifact = loadContractArtifact(CardGameContractArtifactJson as NoirCompiledContract);



/**
 * Type-safe interface for contract CardGame;
 */
export class CardGameContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, CardGameContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, CardGameContract.artifact, wallet) as Promise<CardGameContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, ) {
    return new DeployMethod<CardGameContract>(PublicKeys.default(), wallet, CardGameContractArtifact, CardGameContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, ) {
    return new DeployMethod<CardGameContract>(publicKeys, wallet, CardGameContractArtifact, CardGameContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof CardGameContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<CardGameContract['methods'][M]>
  ) {
    return new DeployMethod<CardGameContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      CardGameContractArtifact,
      CardGameContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return CardGameContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(CardGameContractArtifactJson as NoirCompiledContract);
  }
  

  public static get storage(): ContractStorageLayout<'collections' | 'game_decks' | 'games'> {
      return {
        collections: {
      slot: new Fr(1n),
    },
game_decks: {
      slot: new Fr(2n),
    },
games: {
      slot: new Fr(3n),
    }
      } as ContractStorageLayout<'collections' | 'game_decks' | 'games'>;
    }
    

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** buy_pack(seed: field) */
    buy_pack: ((seed: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** claim_cards(game: integer, cards_fields: array) */
    claim_cards: ((game: (bigint | number), cards_fields: FieldLike[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** join_game(game: integer, cards_fields: array) */
    join_game: ((game: (bigint | number), cards_fields: FieldLike[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** play_card(game: integer, card: struct) */
    play_card: ((game: (bigint | number), card: { strength: (bigint | number), points: (bigint | number) }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** process_message(message_ciphertext: struct, message_context: struct) */
    process_message: ((message_ciphertext: FieldLike[], message_context: { tx_hash: FieldLike, unique_note_hashes_in_tx: FieldLike[], first_nullifier_in_tx: FieldLike, recipient: AztecAddressLike }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** start_game(game: integer) */
    start_game: ((game: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_private_state() */
    sync_private_state: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** view_collection_cards(owner: struct, offset: integer) */
    view_collection_cards: ((owner: AztecAddressLike, offset: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** view_game(game: integer) */
    view_game: ((game: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** view_game_cards(game: integer, player: struct, offset: integer) */
    view_game_cards: ((game: (bigint | number), player: AztecAddressLike, offset: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
}
