// @aztec/noir-contracts.js/src/Claim.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import ClaimContractArtifactJson from '../artifacts/claim_contract-Claim.json' with { type: 'json' };
export const ClaimContractArtifact = loadContractArtifact(ClaimContractArtifactJson as NoirCompiledContract);



/**
 * Type-safe interface for contract Claim;
 */
export class ClaimContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, ClaimContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, ClaimContract.artifact, wallet) as Promise<ClaimContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, target_contract: AztecAddressLike, reward_token: AztecAddressLike) {
    return new DeployMethod<ClaimContract>(PublicKeys.default(), wallet, ClaimContractArtifact, ClaimContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, target_contract: AztecAddressLike, reward_token: AztecAddressLike) {
    return new DeployMethod<ClaimContract>(publicKeys, wallet, ClaimContractArtifact, ClaimContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof ClaimContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<ClaimContract['methods'][M]>
  ) {
    return new DeployMethod<ClaimContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      ClaimContractArtifact,
      ClaimContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return ClaimContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(ClaimContractArtifactJson as NoirCompiledContract);
  }
  

  public static get storage(): ContractStorageLayout<'target_contract' | 'reward_token'> {
      return {
        target_contract: {
      slot: new Fr(1n),
    },
reward_token: {
      slot: new Fr(3n),
    }
      } as ContractStorageLayout<'target_contract' | 'reward_token'>;
    }
    

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** claim(proof_retrieved_note: struct, recipient: struct) */
    claim: ((proof_retrieved_note: { note: { owner: AztecAddressLike, randomness: FieldLike, value: (bigint | number) }, contract_address: AztecAddressLike, metadata: { stage: (bigint | number), maybe_note_nonce: FieldLike } }, recipient: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** constructor(target_contract: struct, reward_token: struct) */
    constructor: ((target_contract: AztecAddressLike, reward_token: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** process_message(message_ciphertext: struct, message_context: struct) */
    process_message: ((message_ciphertext: FieldLike[], message_context: { tx_hash: FieldLike, unique_note_hashes_in_tx: FieldLike[], first_nullifier_in_tx: FieldLike, recipient: AztecAddressLike }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_private_state() */
    sync_private_state: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
}
