// @aztec/noir-contracts.js/src/Crowdfunding.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import CrowdfundingContractArtifactJson from '../artifacts/crowdfunding_contract-Crowdfunding.json' with { type: 'json' };
export const CrowdfundingContractArtifact = loadContractArtifact(CrowdfundingContractArtifactJson as NoirCompiledContract);


      export type WithdrawalProcessed = {
        who: AztecAddressLike
amount: (bigint | number)
      }
    

/**
 * Type-safe interface for contract Crowdfunding;
 */
export class CrowdfundingContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, CrowdfundingContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, CrowdfundingContract.artifact, wallet) as Promise<CrowdfundingContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, donation_token: AztecAddressLike, operator: AztecAddressLike, deadline: (bigint | number)) {
    return new DeployMethod<CrowdfundingContract>(PublicKeys.default(), wallet, CrowdfundingContractArtifact, CrowdfundingContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, donation_token: AztecAddressLike, operator: AztecAddressLike, deadline: (bigint | number)) {
    return new DeployMethod<CrowdfundingContract>(publicKeys, wallet, CrowdfundingContractArtifact, CrowdfundingContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof CrowdfundingContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<CrowdfundingContract['methods'][M]>
  ) {
    return new DeployMethod<CrowdfundingContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      CrowdfundingContractArtifact,
      CrowdfundingContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return CrowdfundingContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(CrowdfundingContractArtifactJson as NoirCompiledContract);
  }
  

  public static get storage(): ContractStorageLayout<'config' | 'donation_receipts'> {
      return {
        config: {
      slot: new Fr(1n),
    },
donation_receipts: {
      slot: new Fr(5n),
    }
      } as ContractStorageLayout<'config' | 'donation_receipts'>;
    }
    

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** donate(amount: integer) */
    donate: ((amount: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** init(donation_token: struct, operator: struct, deadline: integer) */
    init: ((donation_token: AztecAddressLike, operator: AztecAddressLike, deadline: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** process_message(message_ciphertext: struct, message_context: struct) */
    process_message: ((message_ciphertext: FieldLike[], message_context: { tx_hash: FieldLike, unique_note_hashes_in_tx: FieldLike[], first_nullifier_in_tx: FieldLike, recipient: AztecAddressLike }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_private_state() */
    sync_private_state: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** withdraw(amount: integer) */
    withdraw: ((amount: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
    public static get events(): { WithdrawalProcessed: {abiType: AbiType, eventSelector: EventSelector, fieldNames: string[] } } {
    return {
      WithdrawalProcessed: {
        abiType: {
    "kind": "struct",
    "fields": [
        {
            "name": "who",
            "type": {
                "kind": "struct",
                "fields": [
                    {
                        "name": "inner",
                        "type": {
                            "kind": "field"
                        }
                    }
                ],
                "path": "aztec::protocol_types::address::aztec_address::AztecAddress"
            }
        },
        {
            "name": "amount",
            "type": {
                "kind": "integer",
                "sign": "unsigned",
                "width": 128
            }
        }
    ],
    "path": "Crowdfunding::WithdrawalProcessed"
},
        eventSelector: EventSelector.fromString("0x6f9303c1"),
        fieldNames: ["who","amount"],
      }
    };
  }
  
}
