// @aztec/noir-contracts.js/src/Lending.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import LendingContractArtifactJson from '../artifacts/lending_contract-Lending.json' with { type: 'json' };
export const LendingContractArtifact = loadContractArtifact(LendingContractArtifactJson as NoirCompiledContract);



/**
 * Type-safe interface for contract Lending;
 */
export class LendingContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, LendingContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, LendingContract.artifact, wallet) as Promise<LendingContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, ) {
    return new DeployMethod<LendingContract>(PublicKeys.default(), wallet, LendingContractArtifact, LendingContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, ) {
    return new DeployMethod<LendingContract>(publicKeys, wallet, LendingContractArtifact, LendingContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof LendingContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<LendingContract['methods'][M]>
  ) {
    return new DeployMethod<LendingContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      LendingContractArtifact,
      LendingContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return LendingContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(LendingContractArtifactJson as NoirCompiledContract);
  }
  

  public static get storage(): ContractStorageLayout<'collateral_asset' | 'stable_coin' | 'assets' | 'collateral' | 'static_debt'> {
      return {
        collateral_asset: {
      slot: new Fr(1n),
    },
stable_coin: {
      slot: new Fr(2n),
    },
assets: {
      slot: new Fr(3n),
    },
collateral: {
      slot: new Fr(4n),
    },
static_debt: {
      slot: new Fr(5n),
    }
      } as ContractStorageLayout<'collateral_asset' | 'stable_coin' | 'assets' | 'collateral' | 'static_debt'>;
    }
    

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** borrow_private(secret: field, to: struct, amount: integer) */
    borrow_private: ((secret: FieldLike, to: AztecAddressLike, amount: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** borrow_public(to: struct, amount: integer) */
    borrow_public: ((to: AztecAddressLike, amount: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** constructor() */
    constructor: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** deposit_private(from: struct, amount: integer, authwit_nonce: field, secret: field, on_behalf_of: field, collateral_asset: struct) */
    deposit_private: ((from: AztecAddressLike, amount: (bigint | number), authwit_nonce: FieldLike, secret: FieldLike, on_behalf_of: FieldLike, collateral_asset: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** deposit_public(amount: integer, authwit_nonce: field, on_behalf_of: field, collateral_asset: struct) */
    deposit_public: ((amount: (bigint | number), authwit_nonce: FieldLike, on_behalf_of: FieldLike, collateral_asset: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_asset(asset_id: field) */
    get_asset: ((asset_id: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_assets() */
    get_assets: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_position(owner: struct) */
    get_position: ((owner: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** init(oracle: struct, loan_to_value: integer, collateral_asset: struct, stable_coin: struct) */
    init: ((oracle: AztecAddressLike, loan_to_value: (bigint | number), collateral_asset: AztecAddressLike, stable_coin: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** process_message(message_ciphertext: struct, message_context: struct) */
    process_message: ((message_ciphertext: FieldLike[], message_context: { tx_hash: FieldLike, unique_note_hashes_in_tx: FieldLike[], first_nullifier_in_tx: FieldLike, recipient: AztecAddressLike }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** repay_private(from: struct, amount: integer, authwit_nonce: field, secret: field, on_behalf_of: field, stable_coin: struct) */
    repay_private: ((from: AztecAddressLike, amount: (bigint | number), authwit_nonce: FieldLike, secret: FieldLike, on_behalf_of: FieldLike, stable_coin: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** repay_public(amount: integer, authwit_nonce: field, owner: struct, stable_coin: struct) */
    repay_public: ((amount: (bigint | number), authwit_nonce: FieldLike, owner: AztecAddressLike, stable_coin: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_private_state() */
    sync_private_state: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** update_accumulator() */
    update_accumulator: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** withdraw_private(secret: field, to: struct, amount: integer) */
    withdraw_private: ((secret: FieldLike, to: AztecAddressLike, amount: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** withdraw_public(to: struct, amount: integer) */
    withdraw_public: ((to: AztecAddressLike, amount: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
}
