// @aztec/noir-contracts.js/src/NFT.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import NFTContractArtifactJson from '../artifacts/nft_contract-NFT.json' with { type: 'json' };
export const NFTContractArtifact = loadContractArtifact(NFTContractArtifactJson as NoirCompiledContract);



/**
 * Type-safe interface for contract NFT;
 */
export class NFTContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, NFTContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, NFTContract.artifact, wallet) as Promise<NFTContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, admin: AztecAddressLike, name: string, symbol: string) {
    return new DeployMethod<NFTContract>(PublicKeys.default(), wallet, NFTContractArtifact, NFTContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, admin: AztecAddressLike, name: string, symbol: string) {
    return new DeployMethod<NFTContract>(publicKeys, wallet, NFTContractArtifact, NFTContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof NFTContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<NFTContract['methods'][M]>
  ) {
    return new DeployMethod<NFTContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      NFTContractArtifact,
      NFTContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return NFTContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(NFTContractArtifactJson as NoirCompiledContract);
  }
  

  public static get storage(): ContractStorageLayout<'symbol' | 'name' | 'admin' | 'minters' | 'private_nfts' | 'nft_exists' | 'public_owners'> {
      return {
        symbol: {
      slot: new Fr(1n),
    },
name: {
      slot: new Fr(3n),
    },
admin: {
      slot: new Fr(5n),
    },
minters: {
      slot: new Fr(6n),
    },
private_nfts: {
      slot: new Fr(7n),
    },
nft_exists: {
      slot: new Fr(8n),
    },
public_owners: {
      slot: new Fr(9n),
    }
      } as ContractStorageLayout<'symbol' | 'name' | 'admin' | 'minters' | 'private_nfts' | 'nft_exists' | 'public_owners'>;
    }
    

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** cancel_authwit(inner_hash: field) */
    cancel_authwit: ((inner_hash: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** constructor(admin: struct, name: string, symbol: string) */
    constructor: ((admin: AztecAddressLike, name: string, symbol: string) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** finalize_transfer_to_private(token_id: field, partial_note: struct) */
    finalize_transfer_to_private: ((token_id: FieldLike, partial_note: { commitment: FieldLike }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_admin() */
    get_admin: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_private_nfts(owner: struct, page_index: integer) */
    get_private_nfts: ((owner: AztecAddressLike, page_index: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** is_minter(minter: struct) */
    is_minter: ((minter: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** mint(to: struct, token_id: field) */
    mint: ((to: AztecAddressLike, token_id: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** owner_of(token_id: field) */
    owner_of: ((token_id: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** prepare_private_balance_increase(to: struct) */
    prepare_private_balance_increase: ((to: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** private_get_name() */
    private_get_name: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** private_get_symbol() */
    private_get_symbol: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** process_message(message_ciphertext: struct, message_context: struct) */
    process_message: ((message_ciphertext: FieldLike[], message_context: { tx_hash: FieldLike, unique_note_hashes_in_tx: FieldLike[], first_nullifier_in_tx: FieldLike, recipient: AztecAddressLike }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_get_name() */
    public_get_name: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_get_symbol() */
    public_get_symbol: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** set_admin(new_admin: struct) */
    set_admin: ((new_admin: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** set_minter(minter: struct, approve: boolean) */
    set_minter: ((minter: AztecAddressLike, approve: boolean) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_private_state() */
    sync_private_state: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** transfer_in_private(from: struct, to: struct, token_id: field, authwit_nonce: field) */
    transfer_in_private: ((from: AztecAddressLike, to: AztecAddressLike, token_id: FieldLike, authwit_nonce: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** transfer_in_public(from: struct, to: struct, token_id: field, authwit_nonce: field) */
    transfer_in_public: ((from: AztecAddressLike, to: AztecAddressLike, token_id: FieldLike, authwit_nonce: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** transfer_to_private(to: struct, token_id: field) */
    transfer_to_private: ((to: AztecAddressLike, token_id: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** transfer_to_public(from: struct, to: struct, token_id: field, authwit_nonce: field) */
    transfer_to_public: ((from: AztecAddressLike, to: AztecAddressLike, token_id: FieldLike, authwit_nonce: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
}
