// @aztec/noir-contracts.js/src/Orderbook.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import OrderbookContractArtifactJson from '../artifacts/orderbook_contract-Orderbook.json' with { type: 'json' };
export const OrderbookContractArtifact = loadContractArtifact(OrderbookContractArtifactJson as NoirCompiledContract);


      export type OrderFulfilled = {
        order_id: FieldLike
      }
    

      export type OrderCreated = {
        order_id: FieldLike
      }
    

/**
 * Type-safe interface for contract Orderbook;
 */
export class OrderbookContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, OrderbookContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, OrderbookContract.artifact, wallet) as Promise<OrderbookContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, token0: AztecAddressLike, token1: AztecAddressLike) {
    return new DeployMethod<OrderbookContract>(PublicKeys.default(), wallet, OrderbookContractArtifact, OrderbookContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, token0: AztecAddressLike, token1: AztecAddressLike) {
    return new DeployMethod<OrderbookContract>(publicKeys, wallet, OrderbookContractArtifact, OrderbookContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof OrderbookContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<OrderbookContract['methods'][M]>
  ) {
    return new DeployMethod<OrderbookContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      OrderbookContractArtifact,
      OrderbookContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return OrderbookContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(OrderbookContractArtifactJson as NoirCompiledContract);
  }
  

  public static get storage(): ContractStorageLayout<'config' | 'orders'> {
      return {
        config: {
      slot: new Fr(1n),
    },
orders: {
      slot: new Fr(4n),
    }
      } as ContractStorageLayout<'config' | 'orders'>;
    }
    

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** constructor(token0: struct, token1: struct) */
    constructor: ((token0: AztecAddressLike, token1: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** create_order(bid_token: struct, ask_token: struct, bid_amount: integer, ask_amount: integer, authwit_nonce: field) */
    create_order: ((bid_token: AztecAddressLike, ask_token: AztecAddressLike, bid_amount: (bigint | number), ask_amount: (bigint | number), authwit_nonce: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** fulfill_order(order_id: field, authwit_nonce: field) */
    fulfill_order: ((order_id: FieldLike, authwit_nonce: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_order(order_id: field) */
    get_order: ((order_id: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** process_message(message_ciphertext: struct, message_context: struct) */
    process_message: ((message_ciphertext: FieldLike[], message_context: { tx_hash: FieldLike, unique_note_hashes_in_tx: FieldLike[], first_nullifier_in_tx: FieldLike, recipient: AztecAddressLike }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_private_state() */
    sync_private_state: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
    public static get events(): { OrderFulfilled: {abiType: AbiType, eventSelector: EventSelector, fieldNames: string[] }, OrderCreated: {abiType: AbiType, eventSelector: EventSelector, fieldNames: string[] } } {
    return {
      OrderFulfilled: {
        abiType: {
    "kind": "struct",
    "fields": [
        {
            "name": "order_id",
            "type": {
                "kind": "field"
            }
        }
    ],
    "path": "Orderbook::OrderFulfilled"
},
        eventSelector: EventSelector.fromString("0x0e2d8a86"),
        fieldNames: ["order_id"],
      },
OrderCreated: {
        abiType: {
    "kind": "struct",
    "fields": [
        {
            "name": "order_id",
            "type": {
                "kind": "field"
            }
        }
    ],
    "path": "Orderbook::OrderCreated"
},
        eventSelector: EventSelector.fromString("0xc341651c"),
        fieldNames: ["order_id"],
      }
    };
  }
  
}
