// @aztec/noir-contracts.js/src/SimpleToken.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import SimpleTokenContractArtifactJson from '../artifacts/simple_token_contract-SimpleToken.json' with { type: 'json' };
export const SimpleTokenContractArtifact = loadContractArtifact(SimpleTokenContractArtifactJson as NoirCompiledContract);


      export type Transfer = {
        from: AztecAddressLike
to: AztecAddressLike
amount: (bigint | number)
      }
    

/**
 * Type-safe interface for contract SimpleToken;
 */
export class SimpleTokenContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, SimpleTokenContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, SimpleTokenContract.artifact, wallet) as Promise<SimpleTokenContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, name: string, symbol: string, decimals: (bigint | number)) {
    return new DeployMethod<SimpleTokenContract>(PublicKeys.default(), wallet, SimpleTokenContractArtifact, SimpleTokenContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, name: string, symbol: string, decimals: (bigint | number)) {
    return new DeployMethod<SimpleTokenContract>(publicKeys, wallet, SimpleTokenContractArtifact, SimpleTokenContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof SimpleTokenContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<SimpleTokenContract['methods'][M]>
  ) {
    return new DeployMethod<SimpleTokenContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      SimpleTokenContractArtifact,
      SimpleTokenContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return SimpleTokenContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(SimpleTokenContractArtifactJson as NoirCompiledContract);
  }
  

  public static get storage(): ContractStorageLayout<'balances' | 'total_supply' | 'public_balances' | 'symbol' | 'name' | 'decimals'> {
      return {
        balances: {
      slot: new Fr(1n),
    },
total_supply: {
      slot: new Fr(2n),
    },
public_balances: {
      slot: new Fr(3n),
    },
symbol: {
      slot: new Fr(4n),
    },
name: {
      slot: new Fr(6n),
    },
decimals: {
      slot: new Fr(8n),
    }
      } as ContractStorageLayout<'balances' | 'total_supply' | 'public_balances' | 'symbol' | 'name' | 'decimals'>;
    }
    

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** burn_private(from: struct, amount: integer, authwit_nonce: field) */
    burn_private: ((from: AztecAddressLike, amount: (bigint | number), authwit_nonce: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** burn_public(from: struct, amount: integer, authwit_nonce: field) */
    burn_public: ((from: AztecAddressLike, amount: (bigint | number), authwit_nonce: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** cancel_authwit(inner_hash: field) */
    cancel_authwit: ((inner_hash: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** constructor(name: string, symbol: string, decimals: integer) */
    constructor: ((name: string, symbol: string, decimals: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** finalize_mint_to_private(amount: integer, partial_note: struct) */
    finalize_mint_to_private: ((amount: (bigint | number), partial_note: { commitment: FieldLike }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** finalize_transfer_to_private(amount: integer, partial_note: struct) */
    finalize_transfer_to_private: ((amount: (bigint | number), partial_note: { commitment: FieldLike }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** mint_privately(from: struct, to: struct, amount: integer) */
    mint_privately: ((from: AztecAddressLike, to: AztecAddressLike, amount: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** mint_publicly(to: struct, amount: integer) */
    mint_publicly: ((to: AztecAddressLike, amount: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** prepare_private_balance_increase(to: struct, from: struct) */
    prepare_private_balance_increase: ((to: AztecAddressLike, from: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** private_balance_of(owner: struct) */
    private_balance_of: ((owner: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** private_transfer(to: struct, amount: integer) */
    private_transfer: ((to: AztecAddressLike, amount: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** process_message(message_ciphertext: struct, message_context: struct) */
    process_message: ((message_ciphertext: FieldLike[], message_context: { tx_hash: FieldLike, unique_note_hashes_in_tx: FieldLike[], first_nullifier_in_tx: FieldLike, recipient: AztecAddressLike }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_balance_of(owner: struct) */
    public_balance_of: ((owner: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_get_decimals() */
    public_get_decimals: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_get_name() */
    public_get_name: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_get_symbol() */
    public_get_symbol: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_total_supply() */
    public_total_supply: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_transfer(from: struct, to: struct, amount: integer, authwit_nonce: field) */
    public_transfer: ((from: AztecAddressLike, to: AztecAddressLike, amount: (bigint | number), authwit_nonce: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_private_state() */
    sync_private_state: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** transfer_from_private_to_public(from: struct, to: struct, amount: integer, authwit_nonce: field) */
    transfer_from_private_to_public: ((from: AztecAddressLike, to: AztecAddressLike, amount: (bigint | number), authwit_nonce: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** transfer_from_public_to_private(to: struct, amount: integer) */
    transfer_from_public_to_private: ((to: AztecAddressLike, amount: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
    public static get events(): { Transfer: {abiType: AbiType, eventSelector: EventSelector, fieldNames: string[] } } {
    return {
      Transfer: {
        abiType: {
    "kind": "struct",
    "fields": [
        {
            "name": "from",
            "type": {
                "kind": "struct",
                "fields": [
                    {
                        "name": "inner",
                        "type": {
                            "kind": "field"
                        }
                    }
                ],
                "path": "aztec::protocol_types::address::aztec_address::AztecAddress"
            }
        },
        {
            "name": "to",
            "type": {
                "kind": "struct",
                "fields": [
                    {
                        "name": "inner",
                        "type": {
                            "kind": "field"
                        }
                    }
                ],
                "path": "aztec::protocol_types::address::aztec_address::AztecAddress"
            }
        },
        {
            "name": "amount",
            "type": {
                "kind": "integer",
                "sign": "unsigned",
                "width": 128
            }
        }
    ],
    "path": "SimpleToken::Transfer"
},
        eventSelector: EventSelector.fromString("0x70a1894e"),
        fieldNames: ["from","to","amount"],
      }
    };
  }
  
}
