// @aztec/ethereum/dest/test/delayed_tx_utils.js
import { EthAddress } from '@aztec/foundation/eth-address';
import { createLogger } from '@aztec/foundation/log';
import { DateProvider } from '@aztec/foundation/timer';
import { createViemSigner } from '../l1_tx_utils/index.js';
import { L1TxUtilsWithBlobs } from '../l1_tx_utils/l1_tx_utils_with_blobs.js';
import { withDelayer } from './tx_delayer.js';
export class DelayedTxUtils extends L1TxUtilsWithBlobs {
    delayer;
    static fromL1TxUtils(l1TxUtils, ethereumSlotDuration, wallet) {
        const { client, delayer } = withDelayer(wallet, l1TxUtils.dateProvider, {
            ethereumSlotDuration
        });
        const casted = l1TxUtils;
        casted.delayer = delayer;
        casted.client = client;
        return casted;
    }
    enableDelayer(ethereumSlotDuration) {
        const { client, delayer } = withDelayer(this.client, this.dateProvider, {
            ethereumSlotDuration
        });
        this.delayer = delayer;
        this.client = client;
    }
}
export function createDelayedL1TxUtilsFromViemWallet(client, logger = createLogger('L1TxUtils'), dateProvider = new DateProvider(), config, debugMaxGasLimit = false) {
    return new DelayedTxUtils(client, EthAddress.fromString(client.account.address), createViemSigner(client), logger, dateProvider, config, debugMaxGasLimit);
}
