// @aztec/ethereum/dest/contracts/governance_proposer.js
function _ts_decorate(decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for(var i = decorators.length - 1; i >= 0; i--)if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
}
import { memoize } from '@aztec/foundation/decorators';
import { EthAddress } from '@aztec/foundation/eth-address';
import { GovernanceProposerAbi } from '@aztec/l1-artifacts/GovernanceProposerAbi';
import { encodeFunctionData, getContract } from 'viem';
import { encodeSignal, encodeSignalWithSignature, signSignalWithSig } from './empire_base.js';
import { extractProposalIdFromLogs } from './governance.js';
export class GovernanceProposerContract {
    client;
    proposer;
    constructor(client, address){
        this.client = client;
        if (address instanceof EthAddress) {
            address = address.toString();
        }
        this.proposer = getContract({
            address,
            abi: GovernanceProposerAbi,
            client
        });
    }
    get address() {
        return EthAddress.fromString(this.proposer.address);
    }
    async getRollupAddress() {
        return EthAddress.fromString(await this.proposer.read.getInstance());
    }
    async getRegistryAddress() {
        return EthAddress.fromString(await this.proposer.read.REGISTRY());
    }
    getQuorumSize() {
        return this.proposer.read.QUORUM_SIZE();
    }
    getRoundSize() {
        return this.proposer.read.ROUND_SIZE();
    }
    getInstance() {
        return this.proposer.read.getInstance();
    }
    computeRound(slot) {
        return this.proposer.read.computeRound([
            slot
        ]);
    }
    async getRoundInfo(rollupAddress, round) {
        return await this.proposer.read.getRoundData([
            rollupAddress,
            round
        ]);
    }
    getPayloadSignals(rollupAddress, round, payload) {
        return this.proposer.read.signalCount([
            rollupAddress,
            round,
            payload
        ]);
    }
    createSignalRequest(payload) {
        return {
            to: this.address.toString(),
            data: encodeSignal(payload)
        };
    }
    async createSignalRequestWithSignature(payload, slot, chainId, signerAddress, signer) {
        const signature = await signSignalWithSig(signer, payload, slot, await this.getInstance(), this.address.toString(), chainId);
        return {
            to: this.address.toString(),
            data: encodeSignalWithSignature(payload, signature)
        };
    }
    async submitRoundWinner(round, l1TxUtils) {
        const { receipt } = await l1TxUtils.sendAndMonitorTransaction({
            to: this.address.toString(),
            data: encodeFunctionData({
                abi: this.proposer.abi,
                functionName: 'submitRoundWinner',
                args: [
                    round
                ]
            })
        });
        const proposalId = extractProposalIdFromLogs(receipt.logs);
        return {
            receipt,
            proposalId
        };
    }
}
_ts_decorate([
    memoize
], GovernanceProposerContract.prototype, "getRegistryAddress", null);
