// @aztec/ethereum/dest/contracts/inbox.js
import { Buffer16 } from '@aztec/foundation/buffer';
import { EthAddress } from '@aztec/foundation/eth-address';
import { InboxAbi } from '@aztec/l1-artifacts/InboxAbi';
import { getContract } from 'viem';
import { getPublicClient } from '../client.js';
import { checkBlockTag } from './utils.js';
export class InboxContract {
    client;
    inbox;
    static getFromL1ContractsValues(deployL1ContractsValues) {
        const { l1Client, l1ContractAddresses: { inboxAddress } } = deployL1ContractsValues;
        return new InboxContract(l1Client, inboxAddress.toString());
    }
    static getFromConfig(config) {
        const client = getPublicClient(config);
        const address = config.l1Contracts.inboxAddress.toString();
        return new InboxContract(client, address);
    }
    constructor(client, address){
        this.client = client;
        if (address instanceof EthAddress) {
            address = address.toString();
        }
        this.inbox = getContract({
            address,
            abi: InboxAbi,
            client
        });
    }
    get address() {
        return this.inbox.address;
    }
    getContract() {
        return this.inbox;
    }
    async getState(opts = {}) {
        await checkBlockTag(opts.blockNumber, this.client);
        const state = await this.inbox.read.getState(opts);
        return {
            totalMessagesInserted: state.totalMessagesInserted,
            messagesRollingHash: Buffer16.fromString(state.rollingHash),
            treeInProgress: state.inProgress
        };
    }
}
