// @aztec/stdlib/dest/block/block_hash.js
import { Buffer32 } from '@aztec/foundation/buffer';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';
import { schemas } from '../schemas/schemas.js';
/** Hash of an L2 block. */ export class L2BlockHash extends Buffer32 {
    constructor(/** The buffer containing the hash. */ hash){
        super(hash);
    }
    static random() {
        return new L2BlockHash(Fr.random().toBuffer());
    }
    static fromNumber(num) {
        return new L2BlockHash(super.fromNumber(num).toBuffer());
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new L2BlockHash(reader.readBytes(L2BlockHash.SIZE));
    }
    static fromString(str) {
        return new L2BlockHash(super.fromString(str).toBuffer());
    }
    static get schema() {
        return schemas.BufferHex.transform((value)=>new L2BlockHash(value));
    }
    static zero() {
        return new L2BlockHash(Buffer32.ZERO.toBuffer());
    }
    static fromField(hash) {
        return new L2BlockHash(hash.toBuffer());
    }
}
