// @aztec/stdlib/dest/block/validate_block_result.js
import { EthAddress } from '@aztec/foundation/eth-address';
import { schemas } from '@aztec/foundation/schemas';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { z } from 'zod';
import { BlockInfoSchema, deserializeBlockInfo, serializeBlockInfo } from './l2_block_info.js';
import { CommitteeAttestation } from './proposal/committee_attestation.js';
export const ValidateBlockResultSchema = z.union([
    z.object({
        valid: z.literal(true)
    }),
    z.object({
        valid: z.literal(false),
        block: BlockInfoSchema,
        committee: z.array(schemas.EthAddress),
        epoch: schemas.BigInt,
        seed: schemas.BigInt,
        attestors: z.array(schemas.EthAddress),
        attestations: z.array(CommitteeAttestation.schema),
        reason: z.literal('insufficient-attestations')
    }),
    z.object({
        valid: z.literal(false),
        block: BlockInfoSchema,
        committee: z.array(schemas.EthAddress),
        epoch: schemas.BigInt,
        seed: schemas.BigInt,
        attestors: z.array(schemas.EthAddress),
        attestations: z.array(CommitteeAttestation.schema),
        reason: z.literal('invalid-attestation'),
        invalidIndex: z.number()
    })
]);
export function serializeValidateBlockResult(result) {
    if (result.valid) {
        return serializeToBuffer(true);
    }
    const l2Block = serializeBlockInfo(result.block);
    return serializeToBuffer(result.valid, result.reason, l2Block.length, l2Block, result.committee.length, result.committee, result.epoch, result.seed ?? 0n, result.attestors.length, result.attestors, result.attestations.length, result.attestations, result.reason === 'invalid-attestation' ? result.invalidIndex : 0);
}
export function deserializeValidateBlockResult(bufferOrReader) {
    const reader = BufferReader.asReader(bufferOrReader);
    const valid = reader.readBoolean();
    if (valid) {
        return {
            valid
        };
    }
    const reason = reader.readString();
    const block = deserializeBlockInfo(reader.readBuffer());
    const committee = reader.readVector(EthAddress);
    const epoch = reader.readBigInt();
    const seed = reader.readBigInt();
    const attestors = reader.readVector(EthAddress);
    const attestations = reader.readVector(CommitteeAttestation);
    const invalidIndex = reader.readNumber();
    if (reason === 'insufficient-attestations') {
        return {
            valid,
            reason,
            block,
            committee,
            epoch,
            seed,
            attestors,
            attestations: attestations
        };
    } else if (reason === 'invalid-attestation') {
        return {
            valid,
            reason,
            block,
            committee,
            epoch,
            seed,
            attestors,
            invalidIndex,
            attestations: attestations
        };
    } else {
        const _ = reason;
        throw new Error(`Unknown reason: ${reason}`);
    }
}
