// @aztec/ethereum/dest/l1_tx_utils/config.js
import { booleanConfigHelper, floatConfigHelper, getConfigFromMappings, getDefaultConfig, numberConfigHelper } from '@aztec/foundation/config';
export const l1TxUtilsConfigMappings = {
    gasLimitBufferPercentage: {
        description: 'How much to increase calculated gas limit by (percentage)',
        env: 'L1_GAS_LIMIT_BUFFER_PERCENTAGE',
        ...numberConfigHelper(20)
    },
    maxGwei: {
        description: 'Maximum gas price in gwei to be used for transactions.',
        env: 'L1_GAS_PRICE_MAX',
        fallback: [
            'L1_FEE_PER_GAS_GWEI_MAX'
        ],
        ...floatConfigHelper(2000)
    },
    maxBlobGwei: {
        description: 'Maximum blob fee per gas in gwei',
        env: 'L1_BLOB_FEE_PER_GAS_MAX',
        fallback: [
            'L1_BLOB_FEE_PER_GAS_GWEI_MAX'
        ],
        ...floatConfigHelper(3000)
    },
    priorityFeeBumpPercentage: {
        description: 'How much to increase priority fee by each attempt (percentage)',
        env: 'L1_PRIORITY_FEE_BUMP_PERCENTAGE',
        ...numberConfigHelper(20)
    },
    priorityFeeRetryBumpPercentage: {
        description: 'How much to increase priority fee by each retry attempt (percentage)',
        env: 'L1_PRIORITY_FEE_RETRY_BUMP_PERCENTAGE',
        ...numberConfigHelper(50)
    },
    fixedPriorityFeePerGas: {
        description: 'Fixed priority fee per gas in Gwei. Overrides any priority fee bump percentage',
        env: 'L1_FIXED_PRIORITY_FEE_PER_GAS',
        fallback: [
            'L1_FIXED_PRIORITY_FEE_PER_GAS_GWEI'
        ],
        ...floatConfigHelper(0)
    },
    maxSpeedUpAttempts: {
        description: 'Maximum number of speed-up attempts',
        env: 'L1_TX_MONITOR_MAX_ATTEMPTS',
        ...numberConfigHelper(3)
    },
    checkIntervalMs: {
        description: 'How often to check tx status',
        env: 'L1_TX_MONITOR_CHECK_INTERVAL_MS',
        ...numberConfigHelper(1_000)
    },
    stallTimeMs: {
        description: 'How long before considering tx stalled',
        env: 'L1_TX_MONITOR_STALL_TIME_MS',
        ...numberConfigHelper(12_000)
    },
    txTimeoutMs: {
        description: 'How long to wait for a tx to be mined before giving up. Set to 0 to disable.',
        env: 'L1_TX_MONITOR_TX_TIMEOUT_MS',
        ...numberConfigHelper(120_000)
    },
    cancelTxOnTimeout: {
        description: "Whether to attempt to cancel a tx if it's not mined after txTimeoutMs",
        env: 'L1_TX_MONITOR_CANCEL_TX_ON_TIMEOUT',
        ...booleanConfigHelper(true)
    },
    txCancellationFinalTimeoutMs: {
        description: 'How long to wait for a cancellation tx after its last attempt before giving up',
        env: 'L1_TX_MONITOR_TX_CANCELLATION_TIMEOUT_MS',
        ...numberConfigHelper(24 * 12 * 1000)
    },
    txUnseenConsideredDroppedMs: {
        description: 'How long a tx nonce can be unseen in the mempool before considering it dropped',
        env: 'L1_TX_MONITOR_TX_UNSEEN_CONSIDERED_DROPPED_MS',
        ...numberConfigHelper(6 * 12 * 1000)
    }
};
// We abuse the fact that all mappings above have a non null default value and force-type this to Required
export const defaultL1TxUtilsConfig = getDefaultConfig(l1TxUtilsConfigMappings);
export function getL1TxUtilsConfigEnvVars() {
    return getConfigFromMappings(l1TxUtilsConfigMappings);
}
